### ISSUE TITLE
Calendar System / Events

### ISSUE THREAD

**Issue by ThePotatoGuy** [Backdash/MonikaModDev#37]
I've come up with an idea for storing events/dates for the potential calendar system. Each event/date would be represented as a Event object, containing the following:
- title of event
- date/time of event
  - (maybe just date for simplicity)
- event type (date, anniversary, reminder, holiday...)
- event description
- monika's script for this event (or name of label containing monika's script)
- recurrence rules (potential, I feel this is too complex atm)
- event flags (like, did player open game on this day or something)

The events would be stored in persistent as a dict, where the key is the date/time (including year) of the event and the value would be the Event object itself. Past events could be either removed from the dict or left in there depending on whether we want to keep historical data or not.

Monika's scripts for events would have to be pre-defined, which means custom events would not be possible / would be difficult to implement.

That being said, this is really just a brainstorm for event storage, and I would like to hear other's opinions and ideas on this.

**Comment by therationalpi** [comment]
It seems backwards having the date be the key...Shouldn't the events be the key?

**Comment by ThePotatoGuy** [comment]
I was thinking that on a particular day, we could check if its in the persistent, and then pull that day's event.

**Comment by therationalpi** [comment]
I'm thinking this might be the sort of thing where the [Renpy Cookbook Forum](https://lemmasoft.renai.us/forums/viewtopic.php?f=51&t=34131) could yield something.

**Comment by ThePotatoGuy** [comment]
Noted. I'll take a look.

**Comment by polishsnake** [comment]
A calendar system could be complicated later on, if you want to make it very accurate that is.
We have players with many nationalities playing the mod (as well as another issue where Monika asks about player's nationality; hence the idea) and it happens that some nationalities have same events (such as mother's day) at different dates.
I'm not suggesting to look up every single calendar, however some major calendars, or perhaps ask people to make list of events on calendar of their nationality, along with possibly add few more that they know of.
Unless, you want to stick onto the standard English calendar of course (still, British and American calendars do vary, I mean, in England (as well as many more nationalities) there's Christmas and in America there's Thanksgiving).

**Comment by therationalpi** [comment]
The asking about nationality thing will actually be really good here, especially with the current event implementation.

Basically, when we cycle through the events on the calendar on startup, we can also check a few events against nationality. For example, only the US celebrates Thanksgiving, so she can have an event for Thanksgiving, and it's only pushed to the event stack if it's the right day and the player is American.

If we do that, then we can just add *everything* with the correct requirements, and it should work flawlessly.

**Comment by polishsnake** [comment]
It honestly feels nice, to hear that my idea "should work flawlessly".

**Comment by badooga** [comment]
Perhaps Monika should also ask the player when their birthday is - that way, she can wish the player Happy Birthday when that day arrives. If DDLC isn't open on that day, she should say something about how she is sad about not being able to wish the player Happy Birthday on the right date or something.

**Comment by ThePotatoGuy** [comment]
I did a _quick_ browse through the cookbook forum and found two links that might be helpful here:
1. [This one](https://lemmasoft.renai.us/forums/viewtopic.php?f=51&t=27920#p335186) can show a basic calendar screen with appropriate date and time. It doesn't provide total functionality out of the box (no background image), but there's plenty of comments that describe how to do everything.
2. [And this one](https://lemmasoft.renai.us/forums/viewtopic.php?t=22410) shows how to use imagemaps in screens. We could use this for having an interactive/clickable calendar.

**Comment by Kiosho** [comment]
Definitely like the events idea. What if the game logs the date where you first played the mod and confessed your love, and every year Monika brings up your anniversary? Like: "Hey, [Player], today's our anniversary! Did you know that? Gosh, I can't believe it's been (1,2,3... etc.) years."

But what if the player misses their anniversary?
Oh boy...

**Comment by therationalpi** [comment]
@Kiosho That's actually already in the game, but this system will make that more interesting.

**Comment by Danymations** [comment]
I really enjoy the idea of Monika being able to understand the holidays. I joined Github just so I can help with this mod, and the first thing I coded was something about Christmas. I was new to the site though, and I submitted it wrongly haha. I had thought of the idea of her saying “Christmas is near!” but I didn’t know how to code it to only say it when it really is close to Christmas, and to not say it after New Years aha. (I still don’t but i’m trying to learn.)
I went really off topic haha
Basically: I really like the idea of Monika being able to know about current holidays, birthdays, your anniversary, etc. Can’t wait to see how all this turns out and hopefully I can contribute in anyway I can!

**Comment by Kiosho** [comment]
@therationalpi Does MAS mention your anniversary in topics? I'm not finding it.

**Comment by juank713** [comment]
I like it
but I am from costa rica and here the mothers day is diferent thet in other countries
if thers goin to be a calendar for each country  I can help with costa rica one :)

**Comment by antimockingjay** [comment]
If this happens, it may be cool to have the ability to add other people's birthdays. Even if she just makes a little comment. "Oh! I just remembered, it's [name]'s birthday! Did you get [name] a present already? I hope so! When you give it, make sure you tell them I said happy birthday, too!" And that's it. It'd just be super cool to have her reference things really happening, and for a while now I've thought that she'd feel a lot more real if she was able to properly acknowledge the real people I know, so that'd be super cool if that could happen!

**Comment by AnIdiotsavant** [comment]
I'm from the Philippines, maybe I could help with PH calendar, if there's gonna be one

**Comment by therationalpi** [comment]
Alright, so time-based events are now ready for the next release. As for showing those upcoming events to the player with a calendar...that's not implemented yet. But the main functionality for this is done.
